### PR TITLE
lightning module

### PR DESCRIPTION
**Pull request by spirinamayya**
Added lightning module

### CODE REVIEW

File: rectools/models/sasrec.py [hunk 1]
```diff
@@ -600,15 +586,21 @@ def __init__(
...
             item_net_type=item_net_type,
         )
-        self.trainer = Trainer(  # TODO: move to lightning trainer and add option to pass initialized trainer
-            lr=lr,
-            epochs=epochs,
-            device=self.device,
-            loss=loss,
+        self.lightning_model: SasRec
+        self.trainer: Trainer
+        self._trainer = Trainer(
```
Comment (blondered) [inline rectools/models/sasrec.py:591]: Can we pass `logger = verbose > 0` here? To hide all cell outputs and to prevent lightning logging? Please check and add if it will work.

File: rectools/models/sasrec.py [hunk 2]
```diff
@@ -618,10 +610,12 @@ def _fit(
...
         self.model = deepcopy(self._model)  # TODO: check that it works
-        self.model.costruct_item_net(processed_dataset)
+        self.model.construct_item_net(processed_dataset)
+        self.lightning_model = SasRec(self.model, self.lr, self.loss)
+
+        self.trainer = deepcopy(self._trainer)
 
-        self.trainer.fit(self.model, train_dataloader)
-        self.model = self.trainer.model
+        self.trainer.fit(self.lightning_model, train_dataloader)
```
Comment (blondered) [inline rectools/models/sasrec.py:618]: Did you check if `self.model` weights are updated inplace after this?
In recommend you are using `self.lightning_model.torch_model` not `self.model`.
I think it's better to use pure torch net there.

File: rectools/models/sasrec.py [hunk 3]
```diff
@@ -561,6 +475,77 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
...
+        # and will have zeros for every `0` target label
+        loss = self.loss_func(logits.transpose(1, 2), y)  # [batch_size, session_maxlen]
+        loss = loss * w
+        n = (loss > 0).to(loss.dtype)
+        loss = torch.sum(loss) / torch.sum(n)
+        return loss
+
+    def configure_optimizers(self) -> torch.optim.Adam:
+        """TODO"""
+        optimizer = torch.optim.Adam(self.torch_model.parameters(), lr=self.lr, betas=(0.9, 0.98))
```
Comment (blondered) [inline rectools/models/sasrec.py:528]: Let't move this beta value to this class `__init__` default `adam_betas` parameter

**Comment by blondered** [comment]
Add `pos_encoding_type` to SasRecModel init

File: rectools/models/sasrec.py [hunk 4]
```diff
@@ -561,13 +540,86 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
...
+        return self.torch_model(sessions)
+
+    def on_train_start(self) -> None:
+        """TODO"""
+        self._xavier_normal_init()
+
+    def training_step(self, batch: torch.Tensor, batch_idx: int) -> torch.Tensor:
+        """TODO"""
+        x, y, w = batch
+        logits = self.forward(x)  # [batch_size, session_maxlen, n_items + 1]
```
Comment (blondered) [inline rectools/models/sasrec.py:576]: После получения логитов давай будем проверять, какой у нас self.loss. И в зависимости от значения делать соответствующую обработку логитов и расчёт лосса.

**Review by blondered — CHANGES_REQUESTED** [review]

File: rectools/models/sasrec.py [hunk 5]
```diff
@@ -52,6 +53,73 @@ def forward(self, seqs: torch.Tensor, timeline_mask: torch.Tensor, attn_mask: to
...
+    def transform_dataset_u2i(self, dataset: Dataset, users: ExternalIds) -> Dataset:
+        """TODO"""
+        raise NotImplementedError()
+
+    def transform_dataset_i2i(self, dataset: Dataset) -> Dataset:
+        """TODO"""
+        raise NotImplementedError()
+
+
+class SessionEncoderLightningModuleBase(LightningModule):
```
[the remark this answers is not in the document]
Comment (spirinamayya) [reply rectools/models/sasrec.py:111]: fixed

File: rectools/models/sasrec.py [hunk 6]
```diff
@@ -561,13 +540,86 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
...
+
+    def configure_optimizers(self) -> torch.optim.Adam:
+        """TODO"""
+        optimizer = torch.optim.Adam(self.torch_model.parameters(), lr=self.lr, betas=self.adam_betas)
+        return optimizer
+
+    # ce_criterion = torch.nn.CrossEntropyLoss()
+    # https://github.com/NVIDIA/pix2pixHD/issues/9 how could an old bug appear again...
+
+    def _init_loss_func(self, loss: str) -> nn.CrossEntropyLoss:
```
[the remark this answers is not in the document]
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 4 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py:556]: fixed

File: rectools/models/sasrec.py [hunk 7]
```diff
@@ -52,6 +53,73 @@ def forward(self, seqs: torch.Tensor, timeline_mask: torch.Tensor, attn_mask: to
         raise NotImplementedError()
 
 
+class SessionEncoderDataPreparatorBase:
+    """Base class for data preparator. Used only for type hinting."""
+
+    def __init__(
```
[the remark this answers is not in the document]
Comment (spirinamayya) [reply rectools/models/sasrec.py:61]: fixed

File: rectools/models/sasrec.py [hunk 8]
```diff
@@ -561,13 +540,57 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
...
+
+    def on_train_start(self) -> None:
+        """TODO"""
+        self._xavier_normal_init()
+
+    def training_step(self, batch: torch.Tensor, batch_idx: int) -> torch.Tensor:
+        """TODO"""
+        x, y, w = batch
+        logits = self.forward(x)  # [batch_size, session_maxlen, n_items + 1]
+        # We are using CrossEntropyLoss with a multi-dimensional case
```
Comment (blondered) [inline rectools/models/sasrec.py:557]: These comments apply only to softmax loss so we need to place them after `if`

File: rectools/models/sasrec.py [hunk 9]
```diff
@@ -561,13 +540,57 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
...
+
+        # Target label indexes must be passed in a form of [batch_size, session_maxlen]
+        # (`0` index for "PAD" ix excluded from loss)
+
+        # Loss output will have a shape of [batch_size, session_maxlen]
+        # and will have zeros for every `0` target label
+        if self.loss == "softmax":
+            # ce_criterion = torch.nn.CrossEntropyLoss()
+            # https://github.com/NVIDIA/pix2pixHD/issues/9 how could an old bug appear again...
+            loss_func = nn.CrossEntropyLoss(ignore_index=0, reduction="none")
```
Comment (blondered) [inline rectools/models/sasrec.py:570]: Since we are using `loss_func` only to calculate loss can we simple use `torch.nn.functional.cross_entropy`?

File: rectools/models/sasrec.py [hunk 10]
```diff
@@ -561,13 +540,57 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
...
+        #  where n_items + 1 is number of classes
+
+        # Target label indexes must be passed in a form of [batch_size, session_maxlen]
+        # (`0` index for "PAD" ix excluded from loss)
+
+        # Loss output will have a shape of [batch_size, session_maxlen]
+        # and will have zeros for every `0` target label
+        if self.loss == "softmax":
+            # ce_criterion = torch.nn.CrossEntropyLoss()
+            # https://github.com/NVIDIA/pix2pixHD/issues/9 how could an old bug appear again...
```
Comment (blondered) [inline rectools/models/sasrec.py:569]: I don't think we have this bug. Let's drop the comment.

File: rectools/models/sasrec.py [hunk 11]
```diff
@@ -582,14 +605,19 @@ def __init__(
         loss: str = "softmax",
         verbose: int = 0,
         cpu_n_threads: int = 0,
+        deterministic: bool = False,
+        trainer: tp.Optional[Trainer] = None,
         transformer_layers_type: tp.Type[TransformerLayersBase] = SasRecTransformerLayers,  # SASRec authors net
         item_net_type: tp.Type[ItemNetBase] = IdEmbeddingsItemNet,  # item embeddings on ids
+        data_preparator_type: tp.Type[SessionEncoderDataPreparatorBase] = SasRecDataPreparator,
+        lightning_module_type: tp.Type[SessionEncoderLightningModuleBase] = SessionEncoderLightningModule,
+        pos_encoding_type: tp.Type[PositionalEncodingBase] = LearnableInversePositionalEncoding,
```
Comment (blondered) [inline rectools/models/sasrec.py:614]: Let's sort all of the arguments to follow clear logic: all net configuration arguments should be together, then learning configuration args. All of the optional `_type` args should be sorted too. Let's clean this a bit.

**Review by blondered — CHANGES_REQUESTED** [review]

**Comment by blondered** [comment]
Please remove the logger from `logging`

File: rectools/models/sasrec.py [hunk 12]
```diff
@@ -52,6 +55,100 @@ def forward(self, seqs: torch.Tensor, timeline_mask: torch.Tensor, attn_mask: to
         raise NotImplementedError()
 
 
+class SessionEncoderDataPreparatorBase:
```
Comment (blondered) [inline rectools/models/sasrec.py:58]: Please move this class under `# ####  --------------  Data Processor  --------------  #### `

File: rectools/models/sasrec.py [hunk 13]
```diff
@@ -52,6 +55,100 @@ def forward(self, seqs: torch.Tensor, timeline_mask: torch.Tensor, attn_mask: to
...
+    def transform_dataset_u2i(self, dataset: Dataset, users: ExternalIds) -> Dataset:
+        """TODO"""
+        raise NotImplementedError()
+
+    def transform_dataset_i2i(self, dataset: Dataset) -> Dataset:
+        """TODO"""
+        raise NotImplementedError()
+
+
+class SessionEncoderLightningModuleBase(LightningModule):
```
Comment (blondered) [inline rectools/models/sasrec.py:111]: Please move this class under `# ####  --------------  Lightning Model  --------------  #### #`

File: examples/sasrec_metrics_comp.ipynb [hunk 14]
```diff
@@ -987,15 +1040,20 @@
...
     {
      "name": "stderr",
      "output_type": "stream",
      "text": [
       "/data/home/maspirina1/Tasks/RecTools/rectools/dataset/features.py:424: UserWarning: Converting sparse features to dense array may cause MemoryError\n",
-      "  warnings.warn(\"Converting sparse features to dense array may cause MemoryError\")\n"
+      "  warnings.warn(\"Converting sparse features to dense array may cause MemoryError\")\n",
+      "/data/home/maspirina1/Tasks/RecTools/rectools/models/base.py:406: UserWarning: \n",
+      "                Model `<class 'rectools.models.implicit_als.ImplicitALSWrapperModel'>` doesn't support recommendations for cold users,\n",
+      "                but some of given users are cold: they are not in the `dataset.user_id_map`\n",
```
Comment (blondered) [inline examples/sasrec_metrics_comp.ipynb:1054]: Please  check which users are considered cold by ALS. This is an unexpected behaviour and we need to find a reason.

File: rectools/models/sasrec.py [hunk 15]
```diff
@@ -561,13 +540,57 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
...
+
 # ####  --------------  SASRec Model  --------------  #### #
 
 
-class SasRecModel(ModelBase):  # pylint: disable=too-many-instance-attributes
+class SasRecModel(ModelBase):
     """TODO"""
 
-    def __init__(
+    def __init__(  # pylint: disable=too-many-arguments
```
Comment (blondered) [inline rectools/models/sasrec.py:593]: Let's add `dataloader_num_workers` argument and pass it to torch.Dataloader on initialization. This is DSSM example: https://github.com/MobileTeleSystems/RecTools/blob/8a3b7165a6928507a4edfe01ecd100280b7a0a00/rectools/models/dssm.py#L287
We missed this one but it is important.

File: rectools/models/sasrec.py (hunk 8 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 9 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 11 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 12 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 13 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: examples/sasrec_metrics_comp.ipynb (hunk 14 above)
Comment (spirinamayya) [reply examples/sasrec_metrics_comp.ipynb]: Test users passed to ALS were actually cold. In new sasrec_metrics_comp.ipynb I pass all users to SASRec and no cold users to ALS.

File: rectools/models/sasrec.py (hunk 15 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py:598]: fixed

File: rectools/models/sasrec.py [hunk 16]
```diff
@@ -592,26 +597,27 @@ class SasRecModel(ModelBase):
...
-        device: str,
-        n_blocks: int,
-        n_factors: int,
-        n_heads: int,
-        dropout_rate: float,
-        use_pos_emb: bool = True,
+        n_blocks: int = 1,
+        n_heads: int = 1,
+        n_factors: int = 128,
+        use_pos_emb: bool = False,
```
Comment (blondered) [inline rectools/models/sasrec.py:603]: Default should be True because it is a classic part of SASRec

File: rectools/models/sasrec.py [hunk 17]
```diff
@@ -556,40 +503,127 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
...
+    """TODO"""
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        n_blocks: int = 1,
+        n_heads: int = 1,
+        n_factors: int = 128,
+        use_pos_emb: bool = False,
+        dropout_rate: float = 0.2,
+        session_maxlen: int = 32,
```
Comment (blondered) [inline rectools/models/sasrec.py:605]:
```suggestion
        session_max_len: int = 32,
```

**Review by blondered — CHANGES_REQUESTED** [review]

File: rectools/models/sasrec.py [hunk 18]
```diff
@@ -556,40 +503,127 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
...
+            try:
+                torch.nn.init.xavier_normal_(param.data)
+            except ValueError:
+                pass
+
+
+# ####  --------------  SASRec Model  --------------  #### #
+
+
+class SasRecModel(ModelBase):
```
Comment (blondered) [inline rectools/models/sasrec.py:595]:
```suggestion
class SASRecModel(ModelBase):
```
Let's change it now until it's too late :)

File: rectools/models/sasrec.py (hunk 16 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 17 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py (hunk 18 above)
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py [hunk 19]
```diff
@@ -666,8 +660,8 @@ def _recommend_u2i(
 
         ranker = ImplicitRanker(
             self.u2i_dist,
-            user_embs,  # [n_rec_users, n_factors]
-            item_embs_np,  # [n_items + 1, n_factors]
+            user_embs,  # [n_rec_users, factors]
```
[the remark this answers is not in the document]
Comment (spirinamayya) [reply rectools/models/sasrec.py]: fixed

File: rectools/models/sasrec.py [hunk 20]
```diff
@@ -52,6 +51,14 @@ def forward(self, seqs: torch.Tensor, timeline_mask: torch.Tensor, attn_mask: to
         raise NotImplementedError()
 
 
+class PositionalEncodingBase(torch.nn.Module):
+    """Base class for positional encoding. Used only for type hinting."""
```
Comment (feldlime) [inline rectools/models/sasrec.py:55]: If any class is used only for type hinting let's use instead [protocols](https://docs.python.org/3/library/typing.html#typing.Protocol) - it's less confusing and you don't need to implement inheritance and to write such comments

Comment (feldlime) [reply rectools/models/sasrec.py:55]: @blondered FYI

Comment (spirinamayya) [reply rectools/models/sasrec.py]: next pr
